lang/semgrep-grammars/src/semgrep-ruby/src/scanner.cc

```diff
@@ -0,0 +1,1174 @@
+                lexer->result_symbol = SIMPLE_SYMBOL;
+                return true;
+              }
+          }
+
+          return false;
+        }
+        break;
+
+      case '[':
+        // Treat a square bracket as an element reference if either:
+        // * the bracket is not preceded by any whitespace
+        // * an arbitrary expression is not valid at the current position.
+        if (valid_symbols[ELEMENT_REFERENCE_BRACKET] && (!has_leading_whitespace ||
+                                                         !valid_symbols[STRING_START])) {
+          advance(lexer);
+          lexer->result_symbol = ELEMENT_REFERENCE_BRACKET;
+          return true;
+        }
+        break;
+
+      default:
+        break;
+    }
+
+    // Open delimiters for literals
+    if ((valid_symbols[HASH_KEY_SYMBOL] || valid_symbols[IDENTIFIER_SUFFIX]) && (iswalpha(lexer->lookahead) || lexer->lookahead == '_') ||
+        valid_symbols[CONSTANT_SUFFIX] && iswupper(lexer->lookahead)) {
+      TokenType validIdentifierSymbol = iswupper(lexer->lookahead) ? CONSTANT_SUFFIX : IDENTIFIER_SUFFIX;
+      char word[8];
+      int index = 0;
+      while (iswalnum(lexer->lookahead) || lexer->lookahead == '_') {
+        if (index < 8) {
+          word[index] = lexer->lookahead;
+        }
+        index++;
+        advance(lexer);
+      }
+
+      if (valid_symbols[HASH_KEY_SYMBOL] && lexer->lookahead == ':') {
+        lexer->mark_end(lexer);
+        advance(lexer);
+        if (lexer->lookahead != ':') {
+          lexer->result_symbol = HASH_KEY_SYMBOL;
+          return true;
+        }
+      } else if (valid_symbols[validIdentifierSymbol] && lexer->lookahead == '!') {
+        advance(lexer);
+        if (lexer->lookahead != '=') {
+          lexer->result_symbol = validIdentifierSymbol;
+          return true;
+        }
+      }
+
+      return false;
+    }
+
+    // Open delimiters for literals
+    if (valid_symbols[STRING_START]) {
+      Literal literal;
+      literal.nesting_depth = 1;
+
+      if (lexer->lookahead == '<') {
+        advance(lexer);
+        if (lexer->lookahead != '<')
+          return false;
+        advance(lexer);
+
+        Heredoc heredoc;
+        if (lexer->lookahead == '-' || lexer->lookahead == '~') {
+          advance(lexer);
+          heredoc.end_word_indentation_allowed = true;
+        }
+
+        scan_heredoc_word(lexer, &heredoc);
+        if (heredoc.word.empty())
+          return false;
+        open_heredocs.push_back(heredoc);
+        lexer->result_symbol = HEREDOC_START;
+        return true;
+      }
+
+      if (scan_open_delimiter(lexer, literal, valid_symbols)) {
+        literal_stack.push_back(literal);
+        lexer->result_symbol = literal.type;
+        return true;
+      } else {
+        return false;
+      }
+    }
+
+    return false;
+  }
+
+  bool has_leading_whitespace;
+  vector<Literal> literal_stack;
+  vector<Heredoc> open_heredocs;
+};
+
+}  // namespace
+
+extern "C" {
```

Comment:
maybe we should store the change also as a patch, so if anyone has to update scanner.cc and pull the latest, they
can just copy the new one and apply the patch.

Reply:
or maybe we should just store the patch and update 'prep' to do this patching on the original file.